### DISCUSSION TITLE
Add ability to change scope of set/show rules (make them global)

### DISCUSSION THREAD

**Discussion by Andrew15-5** [typst/typst#2153] [Ideas]
In #2127 I was frustrated to admit that I can't `set document(title)` in `title-page()`. And this got me thinking if we can add the ability to modify the scope of specific set/show rules. The basic idea is:

```typ
#let title-page(title) = {
  set-out document(title: title)
  out set document(title: title)
  global set document(title: title)
}
#title-page("custom title")
```

Note that `global` would probably mean that as if it was directly typed in the global scope, while `out` is more of a hoisting 1 level up in the scope levels. This is something I didn't really think through, but both work exactly the same in the example.

This will be an opt-in feature, meaning that this won't break anything. But it also provides to escape the scope of `{}` or `[]`. I haven't thought about how it can affect typesetting workflow (will it change how people write their Typst files). I don't know if this can be abused and/or will this introduce some bugs or incompatibilities and stuff like that. But it seems like a pretty safe and cool feature to me.

**Comment by cnaak** [comment]
The current behavior is indeed quite frustrating. I'm currently writing a book template that is function-based, and the lack of this is really getting in the way...

I'd like to chime in favor of the proposal due to:

- Enabling template developers to write function-based templates;
- Remove the requirement for template initialization: just `import` a template!
- Enable separation between formatting and contents, which is huge!

Imagine being able to apply templates without "formatful" user `show` rules as:

```typst
#import "@preview/some-book-template:0.2.0": *

#book(
  title: "My Book Title",
  author: "Myself, Me",
) // This would operate on the metadata and CHANGE GLOBAL SETTINGS to a "front-matter" style

= Preface
This would be formatted as expected.

= Contents
#outline()
// other "front-matter" stuff, such as list of tables/figures, as in LaTeX

#book-body() // Sets up formatting parameters GLOBALLY and accordingly

= Introduction
Now this is a numbered chapter, etc...

#back-matter() // Sets up formatting parameters GLOBALLY and accordingly

= Acknowledgements
...

= Bibliography
...

#appendix()  // Sets up formatting parameters GLOBALLY and accordingly

= Additional Stuff
Now this is an Appendix
```

Methinks (=It seems to me) the lack of the possibility of template-side — which is where (code, style)-intensive stuff better reside — ability to programatically establish and change `set` and `show` rules globally, prevents the full separation of code and contents. And is really desirable — not only to devs, but general users and writers alike.

**Reply by laurmaedje** [reply]
Linking my response in the other issue here for future readers: https://github.com/typst/typst/issues/93#issuecomment-2301450650

**Reply by Andrew15-5** [reply]
Yes, you should try using global show rule (`#show: ...`).

I created this when I was just starting working in Typst. So I didn't know all the ins and outs of how to style stuff.
